File: src/main/java/com/example/smarthouse/sensor/humidity/HumiditySensor.java
package com.example.smarthouse.sensor.humidity;

import com.example.smarthouse.sensor.DefaultSensor;
import com.example.smarthouse.sensor.config.SensorProtocol;
import com.example.smarthouse.sensor.config.SensorState;
import com.example.smarthouse.sensor.config.SensorType;
import com.example.smarthouse.sensor.message.MessageType;
import com.example.smarthouse.sensor.message.SensorMessage;
import lombok.Data;

import java.util.Date;

public class HumiditySensor extends DefaultSensor {
    private boolean leakDetected;

    public HumiditySensor(int sensorId, String sensorName, SensorProtocol sensorProtocol) {
        super(sensorId, sensorName, SensorState.ON, sensorProtocol, SensorType.HUMIDITY_SENSOR);
        this.leakDetected = false;
    }

    public void detectLeak() {
        leakDetected = true;
        System.out.println("Leak detected");
        activateAlarm(); // Activate alarm when leak is detected
    }

    public void clearLeak() {
        leakDetected = false;
        System.out.println("Leak cleared");
    }

    private void activateAlarm() {
   // Calling an alarm from commands
    }

    @Override
    public SensorMessage getDataFromSensor() {
        return new SensorMessage(
                getSensorId(),
                leakDetected ? 1 : 0,
                new Date(),
                MessageType.SENSOR_DATA,
                leakDetected ? "Leak detected" : "No leak",
                getSensorType()
        );
    }
}

File: src/main/kotlin/com/example/sensor/HumiditySensor.kt
package com.example.smarthouse.sensor.humidity

import com.example.config.SensorProtocol
import com.example.config.SensorState
import com.example.config.SensorType
import com.example.message.MessageType
import com.example.message.Message
import com.example.smarthouse.sensor.DefaultSensor
import java.util.*

data class HumiditySensor(val sensorId: Int, val sensorName: String?, val sensorProtocol: SensorProtocol?) :
    DefaultSensor(sensorId, sensorName, SensorState.ON, sensorProtocol, SensorType.HUMIDITY_SENSOR) {
    private var leakDetected = false
    fun detectLeak() {
        leakDetected = true
        println("Leak detected")
        activateAlarm() // Activate alarm when leak is detected
    }

    fun clearLeak() {
        leakDetected = false
        println("Leak cleared")
    }

    private fun activateAlarm() {
        // Calling an alarm from commands
    }

    override val dataFromSensor: Message
        get() = Message(
            getSensorId(),
            if (leakDetected) 1 else 0,
            Date(),
            MessageType.SENSOR_DATA,
            if (leakDetected) "Leak detected" else "No leak",
            getSensorType()
        )
}
